File: app/src/main/java/com/sardox/timestamper/dialogs/EditNoteDialog.java
package com.sardox.timestamper.dialogs;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.DialogInterface;
import android.support.v7.app.AlertDialog;
import android.view.KeyEvent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.Window;
import android.view.WindowManager;
import android.view.inputmethod.EditorInfo;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.ListView;
import android.widget.TextView;

import com.sardox.timestamper.R;
import com.sardox.timestamper.objects.QuickNote;
import com.sardox.timestamper.objects.Timestamp;
import com.sardox.timestamper.utils.AppSettings;
import com.sardox.timestamper.utils.Consumer;

import java.util.ArrayList;
import java.util.List;

public class EditNoteDialog {

    public EditNoteDialog(final Context context, final Timestamp timestamp, final AppSettings appSettings, final Consumer<String> onNoteEdited) {

        AlertDialog.Builder builder = new AlertDialog.Builder(context);
        final List<String> notesOnly = new ArrayList<>();
        @SuppressLint("InflateParams") View view = LayoutInflater.from(context).inflate(R.layout.text_input_note, null, false);
        final ListView listView = view.findViewById(R.id.quick_notes_listview);
        final LinearLayout quickNotesView = view.findViewById(R.id.quick_notes_view);
        final EditText input = view.findViewById(R.id.input);
        final String timestampNote = timestamp.getNote();
        input.setText(timestampNote);
        input.setSelection(timestampNote.length());
        String defaultString = context.getString(R.string.add_from_widget);
        if (timestampNote.equalsIgnoreCase(defaultString)){
            input.selectAll();
        }
        builder.setView(view);

        if (appSettings.shouldUseQuickNotes()) {
            quickNotesView.setVisibility(View.VISIBLE);
            for (QuickNote q : appSettings.getQuickNotes()) {
                notesOnly.add(q.getNote());
            }
            final ArrayAdapter adapter = new ArrayAdapter<>(context, android.R.layout.simple_list_item_1, notesOnly);
            listView.setAdapter(adapter);
        } else {
            quickNotesView.setVisibility(View.GONE);
        }

        builder.setPositiveButton(android.R.string.ok, new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                dialog.dismiss();
                onNoteEdited.accept(input.getText().toString());
            }
        });

        builder.setNegativeButton(android.R.string.cancel, new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                dialog.cancel();
            }
        });

        final AlertDialog alertDialog = builder.create();
        input.setOnEditorActionListener(new TextView.OnEditorActionListener() {
            @Override
            public boolean onEditorAction(TextView v, int actionId, KeyEvent event) {
                if (actionId == EditorInfo.IME_ACTION_DONE) {
                    onNoteEdited.accept(input.getText().toString());
                    alertDialog.dismiss();
                }
                return false;
            }
        });

        alertDialog.show();

        if (appSettings.shouldShowKeyboardInAddNote()) {
            Window window = alertDialog.getWindow();
            if (window != null) {
                window.setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_STATE_ALWAYS_VISIBLE);
            }
        }

        listView.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
                alertDialog.dismiss();
                onNoteEdited.accept(notesOnly.get(position));
            }
        });
    }


}

File: app/src/main/java/com/sardox/timestamper/dialogs/EditNoteDialog.kt
package com.sardox.timestamper.dialogs

import android.annotation.SuppressLint
import android.content.Context
import android.support.design.chip.Chip
import android.support.design.chip.ChipGroup
import android.support.v7.app.AlertDialog
import android.view.LayoutInflater
import android.view.View
import android.view.inputmethod.EditorInfo
import android.view.inputmethod.InputMethodManager
import android.widget.EditText
import android.widget.LinearLayout
import com.sardox.timestamper.R
import com.sardox.timestamper.objects.Timestamp
import com.sardox.timestamper.utils.AppSettings


class EditNoteDialog(context: Context, timestamp: Timestamp, appSettings: AppSettings, onNoteEdited: (note: String) -> Unit) {

    init {

        val builder = AlertDialog.Builder(context)
        @SuppressLint("InflateParams") val view = LayoutInflater.from(context).inflate(R.layout.text_input_note, null, false)
        val chipGroup = view.findViewById<ChipGroup>(R.id.chipGroup)

        val quickNotesView = view.findViewById<LinearLayout>(R.id.quick_notes_view)
        val input = view.findViewById<EditText>(R.id.input)
        val timestampNote = timestamp.note
        input.setText(timestampNote)
        input.setSelection(timestampNote.length)
        val defaultString = context.getString(R.string.add_from_widget)
        if (timestampNote.equals(defaultString, ignoreCase = true)) {
            input.selectAll()
        }
        builder.setView(view)

        if (appSettings.shouldUseQuickNotes) {
            quickNotesView.visibility = View.VISIBLE
            appSettings.quickNotes.mapNotNull { note -> note.note }.forEach {
                val chip = Chip(chipGroup.context).apply {
                    text = it
                    isClickable = true
                    isCheckable = true
                    setOnCheckedChangeListener { _, isChecked ->
                        when {
                            isChecked -> input.setText(it)
                            else -> if (input.text.toString() == it) input.setText("")
                        }
                    }
                }
                chipGroup.addView(chip)
            }
        } else {
            quickNotesView.visibility = View.GONE
        }

        builder.setPositiveButton(android.R.string.ok) { dialog, _ ->
            dialog.dismiss()
            onNoteEdited.invoke(input.text.toString())
        }

        builder.setNegativeButton(android.R.string.cancel) { dialog, _ -> dialog.cancel() }

        val alertDialog = builder.create()
        input.setOnEditorActionListener { _, actionId, _ ->
            if (actionId == EditorInfo.IME_ACTION_DONE) {
                onNoteEdited.invoke(input.text.toString())
                alertDialog.dismiss()
            }
            false
        }

        alertDialog.show()
        with(input) {
            isFocusable = true
            requestFocus()
        }

        if (appSettings.shouldShowKeyboardInAddNote) {
            val imm = context.getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager?
            imm?.toggleSoftInput(InputMethodManager.SHOW_FORCED, 0)
        }
    }
}